lms/djangoapps/certificates/badge_handler.py

```diff
@@ -0,0 +1,182 @@
+        """
+        Get the name for the badge, based on the course name. Max size is 255.
+        """
+        return _(u"{course_display_name} ({course_mode})").format(
+            course_display_name=course.display_name,
+            # Course modes may be translated, but this whole string won't be available for translation.
+            # pylint: disable=translation-of-non-string
+            course_mode=_(mode),
+        )[:255]
+
+    @staticmethod
+    def badge_description(course):
+        """
+        Returns the badge
+        """
+        if course.end:
+            return _(u"{start_date} to {end_date}").format(
+                start_date=course.start.date(),
+                end_date=course.end.date(),
+            )
+        else:
+            return u"{start_date}".format(course.end.date())
+
+    def create_badge(self, mode):
+        """
+        Create the badge spec for a course's mode.
+        """
+        course = modulestore().get_course(self.course_key)
+        image = BadgeImageConfiguration.image_for_mode(mode)
+        # We don't want to bother validating the file any further than making sure we can detect its MIME type,
+        # for HTTP. The Badgr-Server should tell us if there's anything in particular wrong with it.
+        content_type, __ = mimetypes.guess_type(image.name)
+        if not content_type:
+            raise ValueError(
+                "Could not determine content-type of image! Make sure it is a properly named .png file."
+            )
+        files = {'image': (image.name, image, content_type)}
+        about_path = reverse('about_course', kwargs={'course_id': unicode(self.course_key)})
+        scheme = u"https" if settings.HTTPS == "on" else u"http"
+        data = {
+            'name': self.badge_name_field(course, mode),
+            'criteria': u'{}://{}{}'.format(scheme, settings.SITE_NAME, about_path),
+            'slug': self.course_slug(mode),
+            'description': self.badge_description(course)
+        }
+        result = requests.post(self.badge_create_url, headers=self.get_headers(), data=data, files=files)
+        self.log_if_raised(result, data)
+
+    def create_assertion(self, user, mode):
+        """
+        Register an assertion with the Badgr server for a particular user in a particular course mode for
+        this course.
+        """
+        data = {'email': user.email}
+        response = requests.post(self.assertion_url(mode), headers=self.get_headers(), data=data)
+        self.log_if_raised(response, data)
+        assertion, __ = BadgeAssertion.objects.get_or_create(course_id=self.course_key, user=user)
+        assertion.data = response.json()
+        assertion.save()
+        tracker.emit(
+            'edx.badges.assertion.created', {
+                'user': user.id,
+                'course_id': unicode(self.course_key),
+                'enrollment_mode': mode,
+                'assertion': assertion.data
```

Comment:
After thinking about this more, I'm a little nervous about just dumping the response in here. Primarily because it means we don't control the data that is going into the event, which can cause problems:

1) The size of this payload increases dramatically.
2) The remote service changes the payload to include data that we do not want to save in our tracking logs.
3) The remote service changes the format and/or structure of the data which, in turn, breaks the code of researchers who are analyzing the log and making assumptions about the format and structure.

I think I would feel better about it if we explicitly copied the data we want from the response into the event.